### ISSUE TITLE
Rotary encoder [QUESTION]

### ISSUE THREAD

**Issue by lukaskloepping** [pideck/pideck#42]
Is it possible to embed a rotary encoder, just like pioneer did it next to the cdj displays?
Rotate the knob - scroll through your tracks
Press the knob - load the track
?
Would be a great feature for me instead of using a crappy touch screen! :)

**Comment by danielhjames** [comment]
Yes, this would be a great enhancement! The rotate action would be mapped to the up/down arrow keys and the press action to the F1 key (for the first deck). If you wanted two turntables or CDJs per PiDeck it would be a little more complicated. Prototypes welcome :-)

**Comment by danielhjames** [comment]
See http://guy.carpenter.id.au/gaugette/2013/01/14/rotary-encoder-library-for-the-raspberry-pi/ for some ideas.

**Comment by lukaskloepping** [comment]
Cool!
I will start the project as soon as possible :-D
I'd like to create a seperate pi setup for each one of my two technics to avoid performance problems, so this should be pretty simple :)
The plan was to create a case where my mixer, the two pi's, two displays, two knobs, two soundcards, power supplys etc fit in, so i can reach a standalone DVS setup i only need to connect to my turntables and my studio monitoring setup
With the knobs and maybe a "back" button this should work as good as a cdj and its ultra portable too :) 👍

**Comment by danielhjames** [comment]
If you are using two Pi's and two displays you might like to consider two cases, as this will make it easier to swap out one unit and carry on should anything go wrong with one of the systems. Please send pictures or video of your project :-)

**Comment by BennPaul** [comment]
For audioinjector this may be helpful:
http://www.flatmax.org/phpbb/viewtopic.php?f=5&t=4

My rotary encoder is on the way. I'll try to get it running as soon as it arrives.

**Comment by BennPaul** [comment]
First version for selecting tracks via a rotary encoder.

script: **rotary_track_select.py**

For the circuit of the rotary encoder see:

https://i0.wp.com/henrysbench.capnfatz.com/wp-content/uploads/2015/05/Keyes-KY-040-Rotary-Encoder-Module-Schematic.png?resize=300%2C183

The resistors are 10 k Ohm each. Voltage is 3.3 V

CLK connected to GPIO06
DT connected to GPIO05
SW connected to GIPO13

None of these pins is used by the audioinjector card.

You need the following three components to run the script:

1. xdotool
2. RPi.GPIO for python
3. KY040 for python

for xdotool:
`sudo apt-get install xdotool`

for RPi.GPIO follow the installation instructions on: 
https://sourceforge.net/p/raspberry-gpio-python/wiki/install/

for KY040 see:
https://github.com/conradstorz/KY040


If it is not working with the AudioInjector card:
I tried it for ages with AI plugged to the pi without any success. So I had the chance to try a lot of different tutorials on the internet searching for my mistake ;-) After I had no more idea left than unplugging the AI card, it worked. Although all other GPIO stuff I tried with the AI card worked but not the rotary encoder. With a 2nd AI card I hadn't any problems. 

Keep this in mind if you have problems using the script.

Annotation:
Maybe there are better time intervals for the sleep command. (cpu load vs. rotating smoothly)

**Comment by danielhjames** [comment]
Awesome, thanks! We'll take a closer look at this.

**Comment by obbardc** [comment]
Hi BennPaul,

Thanks for this.  I am a bit weary of actually accepting this as a pull request yet, as we will have to package up both RPi.GPIO and KY040 for Debian.

**Comment by BennPaul** [comment]
Hey,

the code needed from KY040 is around 40 lines and could be easily integrated in the script itself.
I didn't do it because I'm not familiar with licence stuff. 
The KY040 licence says: https://github.com/conradstorz/KY040/blob/master/LICENSE
Is it working together with pideck?

The RPi.GPIO licence says: https://sourceforge.net/p/raspberry-gpio-python/code/ci/default/tree/LICENCE.txt
But I think the integration is more complicated than coping and pasting 40 lines.

**Comment by danielhjames** [comment]
Hi Ben, yes, these licenses are compatible with the GNU GPL software we use, please see https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses

There are armhf debs available at https://sourceforge.net/projects/raspberry-gpio-python/files/raspbian-jessie/ and the KY040 code is so small we could incorporate it into the pideck package, I think.

**Comment by BennPaul** [comment]
Sorry for the late answer, a lot of work at the moment. I'll try to update the script on the weekend.

**Comment by BennPaul** [comment]
PR with new script

**Comment by moosair** [comment]
Hi, I like the idea with the KY040 very much. But i have troubles to start the script. 
import: command not found
class: command not found
the same with CLOCKWISE and ANTICLOCKWISE and DEBOUNCE
BennPaul, how do you mean:" With a 2nd AI card I hadn't any problems." did you put 2 AI on one PI? or did you try it with another different AI?
Thank you and greetings  andy

**Comment by uroborov** [comment]
Hi! that is quite interesting, i think i will give it a ry on my project for a standalone pideck. do you think it can be easy to add a switch button for the bypass function? how should they connect to the GPIOs?

**Comment by danielhjames** [comment]
Hi @uroborov the bypass button simply runs some amixer commands, please see lines 79-87 of https://github.com/pideck/pideck/blob/master/touchscreen_controls_L

So you would need a program to listen for the button presses and run similar commands, depending on your soundcard hardware. Try http://raspi.tv/2013/how-to-use-interrupts-with-python-on-the-raspberry-pi-and-rpi-gpio

**Comment by interym** [comment]
it seems like adafruit has created a module for retrogame emulators which is pretty convinient. just wanted to share since i remembered browsing this issue once. will try it some time https://github.com/adafruit/Adafruit-Retrogame

**Comment by interym** [comment]
so i cannot install since i have problems to download the programm. there is neither curl installed nor can i install curl via apt-get (seems i cannot install anything with apt get). whats the reason for this, is there a way around or does anybody have an idea how to download the whole programm and install from usb?

**Comment by danielhjames** [comment]
Hi @interym you will need to ensure the PiDeck has a valid apt source configured before you can install any extra software. By default this is not configured, as I recall.

**Comment by interym** [comment]
hi, turns out retrogame uses alot of libraries which are not part of pideck. i was able to get apt working but all the repos in apt and curl are not signed i cant fix this propperly so no easy solution here...
